android/src/main/cpp/JNIRegistry.cpp

```diff
@@ -0,0 +1,58 @@
+#include "JNIRegistry.h"
+#include <memory>
+
+JNIRegistry::JNIRegistry(JNIEnv* env) {
+    this->env = env;
+}
+
+std::string JNIRegistry::generateMethodKey(
+        std::string className,
+        std::string methodName,
+        std::string methodSignature) const {
+    return className + "(" + methodName + "(" + methodSignature + "))";
+}
+
+std::tuple<jclass, jmethodID> JNIRegistry::getClassAndMethod(
+        std::string className,
+        std::string methodName,
+        std::string methodSignature,
+        JNIMethodMode methodMode,
+        JNIEnv *currentEnv) {
+    currentEnv = (currentEnv == nullptr) ? this->env : currentEnv;
+    jclass classPtr = nullptr;
+    jmethodID methodPtr = nullptr;
+    // check if class is already stored, if not, try to obtain it
+    auto classIt = this->classes.find(className);
+    if (classIt != this->classes.end()) { // found
+        classPtr = classIt->second;
+    } else { // not found
+        classPtr = currentEnv->FindClass(className.c_str());
+        // register obtained class
+        classPtr = (jclass)currentEnv->NewGlobalRef(classPtr);
```

Comment:
I think we also have to delete this ref somewhere.

Reply:
fixed ac5c74f